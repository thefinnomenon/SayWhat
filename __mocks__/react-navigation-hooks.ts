const navigate = jest.fn();

const useNavigation = () => {
  return {
    navigate,
  };
};

type useNavigationParamProps = {
  value: string;
  default?: string;
};

const useNavigationParam = (props: useNavigationParamProps) => {
  return props.value;
};

export { useNavigation, useNavigationParam, navigate };
