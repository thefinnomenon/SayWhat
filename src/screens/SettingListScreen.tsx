import React, { useState } from 'react';
import { ScrollView } from 'react-native';
import { Theme } from '../../types';
import styled from 'styled-components/native';
import { useNavigationParam } from 'react-navigation-hooks';
import ListRow from '../components/ListRow';

type ListOptions = {
  key: string;
  value: string;
};

export const SettingListScreen = () => {
  const options = useNavigationParam('options');
  const selectedKey = useNavigationParam('selectedKey');
  const onSelect = useNavigationParam('onSelect');

  const [selectedRow, setSelectedRow] = useState(selectedKey);

  return (
    <Container>
      <ScrollView>
        {options.map(({ key, value }: ListOptions) => (
          <ListRow
            key={key}
            id={key}
            value={value}
            isSelected={selectedRow === key}
            onSelect={val => {
              onSelect(val);
              setSelectedRow(key);
            }}
          />
        ))}
      </ScrollView>
    </Container>
  );
};

// @ts-ignore: react-navigation has messed up types
SettingListScreen.navigationOptions = ({ navigation }) => ({
  title: navigation.getParam('title', 'Title'),
});

const Container = styled.View<Theme>`
  flex: 1;
  background: ${props => props.theme.colors.background};
`;

export default SettingListScreen;
