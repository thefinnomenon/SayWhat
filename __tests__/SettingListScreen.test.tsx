import 'react-native';
import React from 'react';
import { render, fireEvent } from 'react-native-testing-library';
import 'jest-styled-components';
import ThemeManager from '../contexts/ManageThemeContext';
import { I18nextProvider } from 'react-i18next';
import i18 from '../i18n';
import 'jest-styled-components';

import SettingListScreen from '../src/screens/SettingListScreen';
import { useNavigationParam } from 'react-navigation-hooks';

jest.mock('react-navigation-hooks', () => {
  return {
    useNavigationParam: jest.fn().mockImplementation(param => {
      switch (param) {
        case 'options':
          return [
            { key: '1', value: 'One' },
            { key: '2', value: 'Two' },
            { key: '3', value: 'Three' },
          ];
        case 'selectedKey':
          return '2';
        case 'onSelect':
          return jest.fn();
      }
    }),
  };
});

const renderComponent = () =>
  render(
    <ThemeManager>
      <I18nextProvider i18n={i18}>
        <SettingListScreen />
      </I18nextProvider>
    </ThemeManager>,
  );

it('renders correctly', () => {
  const tree = renderComponent().toJSON();
  expect(tree).toMatchSnapshot();
});

test('selecting a row triggers onSelect with the corresponding row value.', () => {
  const { getByText } = renderComponent();
  fireEvent.press(getByText('One'));
  // TODO: How to check that the onSelect function fired
  //expect(onSelect).toBeCalledWith('One');
});

test('selecting a row sets that row as selected', () => {
  const { getByText } = renderComponent();
  fireEvent.press(getByText('One'));
  console.log(getByText('One'));
  // TODO: How to check if the row is now selected
  //expect(getByText('One')).toHaveProperty('pendingProps.isSelected: true');
});
