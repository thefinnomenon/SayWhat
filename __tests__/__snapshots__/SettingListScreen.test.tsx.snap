// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
>
  <RCTScrollView>
    <View>
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 46,
            "justifyContent": "space-between",
            "opacity": 1,
          }
        }
      >
        <Text
          isSelected={false}
          style={
            Array [
              Object {
                "color": "#202020",
                "fontSize": 18,
                "fontWeight": "normal",
                "marginLeft": 8,
              },
            ]
          }
        >
          One
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#bdbdbd",
              "height": 1,
              "marginBottom": "auto",
              "marginLeft": "auto",
              "marginRight": "auto",
              "marginTop": "auto",
              "width": "95%",
            },
          ]
        }
      />
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 46,
            "justifyContent": "space-between",
            "opacity": 1,
          }
        }
      >
        <Text
          isSelected={true}
          style={
            Array [
              Object {
                "color": "#202020",
                "fontSize": 18,
                "fontWeight": "bold",
                "marginLeft": 8,
              },
            ]
          }
        >
          Two
        </Text>
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": undefined,
                "fontSize": 12,
              },
              Array [
                Object {
                  "color": "#202020",
                  "fontSize": 20,
                  "marginRight": 12,
                },
              ],
              Object {
                "fontFamily": "FontAwesome5Free-Regular",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {
                "fontWeight": "400",
              },
            ]
          }
        >
          ÔÅò
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#bdbdbd",
              "height": 1,
              "marginBottom": "auto",
              "marginLeft": "auto",
              "marginRight": "auto",
              "marginTop": "auto",
              "width": "95%",
            },
          ]
        }
      />
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 46,
            "justifyContent": "space-between",
            "opacity": 1,
          }
        }
      >
        <Text
          isSelected={false}
          style={
            Array [
              Object {
                "color": "#202020",
                "fontSize": 18,
                "fontWeight": "normal",
                "marginLeft": 8,
              },
            ]
          }
        >
          Three
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#bdbdbd",
              "height": 1,
              "marginBottom": "auto",
              "marginLeft": "auto",
              "marginRight": "auto",
              "marginTop": "auto",
              "width": "95%",
            },
          ]
        }
      />
    </View>
  </RCTScrollView>
</View>
`;
